File: cpp/src/io/viz.cpp
#include "io/viz.hpp"

#include <sstream>

#include <opencv2/imgproc.hpp>

#include "config.hpp"

namespace lapi {

cv::Mat draw(const cv::Mat& frame, const std::vector<Detection>& detections,
             int frame_count) {
    cv::Mat out = frame.clone();
    for (const auto& det : detections) {
        const cv::Scalar color = config::COLORS[0];
        const cv::Point p1(det.box.x, det.box.y);
        const cv::Point p2(det.box.x + det.box.width, det.box.y + det.box.height);

        std::ostringstream label;
        label << " " << det.score << "  " << det.text;

        // Masque translucide à l'intérieur de la boîte
        if (!det.mask.empty()) {
            cv::Mat overlay = out.clone();
            overlay(det.box).setTo(color, det.mask == 1);
            cv::addWeighted(out, 0.6, overlay, 0.4, 0, out);
        }

        std::vector<cv::Point> poly;
        for (const auto& pt : det.polygon) poly.emplace_back(pt);
        cv::polylines(out, poly, /*isClosed=*/true, color, 2);
        cv::rectangle(out, p1, p2, color, 2);

        // Cartouche du label au-dessus de la boîte
        int baseline = 0;
        const cv::Size ts = cv::getTextSize(label.str(), cv::FONT_HERSHEY_SIMPLEX,
                                            0.6, 2, &baseline);
        cv::rectangle(out, cv::Point(p1.x, p1.y - ts.height - 10),
                      cv::Point(p1.x + ts.width, p1.y), color, -1);
        cv::putText(out, label.str(), cv::Point(p1.x, p1.y - 5),
                    cv::FONT_HERSHEY_SIMPLEX, 0.6, cv::Scalar(255, 255, 255), 2);
    }

    cv::putText(out, "frame " + std::to_string(frame_count), cv::Point(10, 30),
                cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(0, 255, 0), 2);
    return out;
}

}  // namespace lapi

File: src/io/viz.py
"""Annotation visuelle des frames : masques, polygones, boîtes et labels."""
from __future__ import annotations

import cv2
import numpy as np

from src.config import COLORS


def draw(frame, detections: list, frame_count: int):
    """Dessine les détections (masque translucide, polygone, boîte, score + texte)."""
    out = frame.copy()
    for det in detections:
        color = COLORS[0]
        x1, y1, x2, y2 = det["box"]
        label = f" {det['score']}  {det.get('text', '')}"

        # Masque translucide à l'intérieur de la boîte
        overlay = out.copy()
        overlay[y1:y2, x1:x2][det["mask"] == 1] = color
        out = cv2.addWeighted(out, 0.6, overlay, 0.4, 0)

        cv2.polylines(out, [np.array(det["polygon"], dtype=np.int32)],
                      isClosed=True, color=color, thickness=2)
        cv2.rectangle(out, (x1, y1), (x2, y2), color, 2)

        # Cartouche du label au-dessus de la boîte
        (tw, th), _ = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.6, 2)
        cv2.rectangle(out, (x1, y1 - th - 10), (x1 + tw, y1), color, -1)
        cv2.putText(out, label, (x1, y1 - 5), cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 2)

    cv2.putText(out, f"frame {frame_count}", (10, 30),
                cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)
    return out
